include/zephyr/kernel.h

```diff
@@ -6076,6 +6076,40 @@ static inline void k_cpu_atomic_idle(unsigned int key)
  */
 #define k_panic()	z_except_reason(K_ERR_KERNEL_PANIC)
 
+#if !__ASSERT_ON
+#if defined(CONFIG_BUG_NO_PANIC)
+#define __BUG_PANIC()
+#else
+#define __BUG_PANIC() k_panic()
+#endif
+
+#if defined(CONFIG_BUG_PRINT_LOCATION)
+#define __BUG_PRINT(fmt, ...) printk(fmt, ##__VA_ARGS__)
+#else
+#define __BUG_PRINT(fmt, ...)
+#endif
+#define __BUG_LOC() __BUG_PRINT("BUG @ %s:%d\n", __FILE__, __LINE__)
+
+#define __BUG()                                                                                    \
+	do {                                                                                       \
+		__BUG_LOC();                                                                       \
+		__BUG_PANIC();                                                                     \
+	} while (false)
+
+#define __BUG_ON(test)                                                                             \
+	do {                                                                                       \
+		if (!(test)) {                                                                     \
+			__BUG_LOC();                                                               \
+			__BUG_PANIC();                                                             \
+		}                                                                                  \
+	} while (false)
+
+#else
+/* If asserts are enabled, then these behave exactly like asserts. */
+#define __BUG()        __ASSERT_NO_MSG(false)
+#define __BUG_ON(test) __ASSERT_NO_MSG(test)
```

Comment:
It would be good to add doxygen (and maybe a doxygen group) so that it appears in the right place on docs.zephyrproject.org .

Looks good to me though. Thanks for putting it together.

I'll hold off on approving until some others have reviewed. Maybe @peter-mitsis or @nashif?

Reply:
done